### PR TITLE
Adding missing defaults to the --help output

### PR DESCRIPTION
**Pull request by tharanga**
Issue: #390

### CODE REVIEW

**Comment by dormando** [comment]
The defaults shouldn't be in two places; need to move to define's and then be used in the help section, else they're either going to drift or just be more work to maintain.

A few exceptions, such as ext_item_age, should be listed in a human readable format, ie "no age limit" vs INT_MAX

**Comment by tharanga** [comment]
> The defaults shouldn't be in two places; need to move to define's and then be used in the help section, else they're either going to drift or just be more work to maintain.
I started with replacing all defaults in the help message with their default settings values, so changing values will not make the help message out of sync. However, there are some instances more help text has added based on the default value; hence abandoned that approach. Not sure adding asserts along with the help message is a good idea for those instances. We could also do the best effort and leave the rest like ext_item_age. 
> 
> A few exceptions, such as ext_item_age, should be listed in a human readable format, ie "no age limit" vs INT_MAX

**Comment by dormando** [comment]
Not sure I follow what you mean by "more help text has added based on the default value"

**Comment by tharanga** [comment]
For example:
- lru_crawler_tocrawl: max items to crawl per slab per run
                          default is 0 (unlimited)
Default value has a special meaning in this context. I initially modified this with an assert, so anyone who changes the default value will be forced to change the description, not sure whether that's a good idea though.

File: memcached.c [hunk 1]
```diff
@@ -6419,78 +6419,91 @@ static void usage(void) {
...
+           "   - ext_threads:         number of IO threads to run. (default: 1)\n"
+           "   - ext_item_size:       store items larger than this (bytes; default: 512)\n"
+           "   - ext_item_age:        store items idle at least this long (default: 4294967295)\n"
+           "   - ext_low_ttl:         consider TTLs lower than this specially (default: 0)\n"
            "   - ext_drop_unread:     don't re-write unread values during compaction\n"
-           "   - ext_recache_rate:    recache an item every N accesses\n"
+           "                          (default: disabled)\n"
+           "   - ext_recache_rate:    recache an item every N accesses (default: 2000)\n"
            "   - ext_compact_under:   compact when fewer than this many free pages\n"
+           "                          (default: total storage / (ext_page_size * 4) )\n"
```
Comment (tharanga) [inline memcached.c]: Is it alright to be descriptive for a default like this? There's a danger that this could go out of sync with the actual logic.

**Comment by dormando** [comment]
> For example:
> 
> * lru_crawler_tocrawl: max items to crawl per slab per run
>   default is 0 (unlimited)
>   Default value has a special meaning in this context. I initially modified this with an assert, so anyone who changes the default value will be forced to change the description, not sure whether that's a good idea though.

No that's like the item_age one. Should just be a handful of exceptions like that though?

File: memcached.c (hunk 1 above)
Comment (dormando) [reply memcached.c]: I think so, though maybe "1/4th of the assigned storage" is more clear.

**Comment by tharanga** [comment]
Yes, it is like item_age and there's only a few of them. Let me change others.

**Comment by tharanga** [comment]
I changed the approach and almost finished changes. Let me revive the branch and update the PR.

**Comment by dormando** [comment]
This might get nuked by an upcoming code reorg, in case you can find some time to wrap it up

File: memcached.c [hunk 2]
```diff
@@ -7601,76 +7608,104 @@ static void usage(void) {
...
-           "                          table should be. normally grows at runtime.\n"
+           "                          table should be. normally grows at runtime. (default starts at: %d)\n"
            "                          set based on \"STAT hash_power_level\"\n"
            "   - tail_repair_time:    time in seconds for how long to wait before\n"
            "                          forcefully killing LRU tail item.\n"
            "                          disabled by default; very dangerous option.\n"
            "   - hash_algorithm:      the hash table algorithm\n"
            "                          default is murmur3 hash. options: jenkins, murmur3\n"
-           "   - lru_crawler:         enable LRU Crawler background thread\n"
+           "   - lru_crawler:         enable LRU Crawler background thread \n"
```
Comment (tharanga) [inline memcached.c]: lru crawler and maintainer threads are enabled by default.

**Comment by dormando** [comment]
not sure if you mean for me to check this yet :P please say so clearly when ready.

**Comment by tharanga** [comment]
It's ready to review. I might be missing a couple things, but appreciate if you can verify the approach (e.g. asserts for "explained defaults").

File: memcached.c (hunk 2 above)
Comment (dormando) [reply memcached.c]: Most (all?) boolean options like this have a "no_" counterpart. could just change this to "no_lru_crawler: disable etc"

File: memcached.c [hunk 3]
```diff
@@ -8422,6 +8457,8 @@ int main (int argc, char **argv) {
 
     /* init settings */
     settings_init();
+    assert(hash_type == MURMUR3_HASH);// if this assertion fails, change the default hash_algorithm
```
Comment (dormando) [inline memcached.c]: why is this here when the other asserts are up in the help text?

related; do any tests actually call -h explicitly? A bunch of the supports_* stuff does, but perhaps this could use a singular test for "do the help asserts fire"?

**Comment by dormando** [comment]
looks pretty good I think. added a couple comments, and I think there's an errand formatting issue here or there.

I think the asserts are okay, just make sure they're close to each other and maybe add the explicit test. probably worth doing it in 00-startup so we don't get mysterious test failures later on.

File: memcached.c (hunk 3 above)
Comment (tharanga) [reply memcached.c]: `hash_type` is local to the main method, it's not a part of the settings unless we refactor it.
+1 on a test, I will add one.

**Comment by dormando** [comment]
Neat. looks like it makes sense. anything else you want to do here?

**Comment by tharanga** [comment]
Nope. I will squash if all is good.

**Comment by dormando** [comment]
go ahead and squash. I'll make a final review later and fix anything I've missed myself.

**Comment by dormando** [comment]
shoot. was doing one last review and started fixing some things, but there's one big pattern that I'd like you to fix:

finally realized a bunch of spots have text like: `(default: 64mb)` or `(default: 8kb)`, etc. but the actual commandline arguments don't take numbers like that, which will confuse users. There were some existing cases but it was separated by a space so if a user input `-m 64 MB` they'd get an error or it wouldn't do anything, but they might try to input `-m 8g` which currently starts (ugh) without any warning (ugh), but gives you 8 megabytes of RAM.

`-I` can take suffixes of 'm', 'k', etc, but not kb/mb/etc. looks like a one character type in the original text too.

in all cases the text says "size in kilobytes/megabytes/etc where etc", so just change the `(default: 8kb)` and similar to `(default: 8)`.

Thanks!

**Comment by dormando** [comment]
@tharanga wanna give this one last set of edits? :) going to be ripping apart memcached.c shortly. want to get this in ahead of it.

**Comment by tharanga** [comment]
@dormando we can squash if this looks good.

**Comment by dormando** [comment]
@tharanga looks good! go ahead and squash. I'll merge it soon.

**Comment by dormando** [comment]
out in 1.5.21, thanks!
